### DISCUSSION TITLE
Why the app does not run in the background but only when it's active in the foreground? 😐

### DISCUSSION THREAD

**Discussion by gab81** [BimmerGestalt/AAIdrive#395] [Ideas]
hello,

I love the app and use it very often to get extra data such as oil temp, engine temp and liters left etc.

One key functionality we recently lost in the main mini app is saving the location of the car - the new mini app does save the location of the car BUT when you see it, you cannot click and go to maps, which is super essential and useful if let's say you want to walk back to the car, etc. old app had that.

Now, the Android Auto Drive DOES have that but one key element that either does not work or could be improved is that the app only saves the location of the car when you park if it's LOADED and in the foreground - sure you can click on it one sec but you have to "remember" doing it, otherwise it won't save and it still show the location of where i was at home hours earlier.

For example, if you are using gmaps and earlier you peeked at the app, hey engine temp is great - let's move on or i have 23 liters left, GREAT, let's move on and then carry on using gmaps and then arrive to destination then the app does NOT save the location.

Is it possible to have it run in the background (i checked all options and I have "allow app to run in background", all connectivity works) so that it will run in parallel with other apps at the same time and automatically tracking the last vehicle location? or if it cannot do it in real time all the time how about every 5 minutes? that would be an improvement already.

I am using this version: 1.3.28 Jul 2021 05.54.00

many thanks,
Gabrio

**Comment by hufman** [comment]
I'm glad you like this feature!

The background updating you mention should already be in place, and I just went out to test and confirm that I didn't accidentally add a bug recently :) My goal is to provide full functionality without even needing to take the phone out of the pocket at all, and I just confirmed that (at least for me) it was able to update the window, sunroof, and position status without needing to take my phone out of my pocket or open the app at all.

In fact, updating the car information is relatively tightly integrated into the app flow: The same [CarInformationUpdater instance](https://github.com/BimmerGestalt/AAIdrive/blob/48f7f2c9b7813754b838f1bbbea294c0558cd7c1/app/src/main/java/me/hufman/androidautoidrive/CarInformation.kt#L118) that is responsible for updating the [Cached CDS info](https://github.com/BimmerGestalt/AAIdrive/blob/00bc2fbc88d2784391118a29153865e9973bc484/app/src/main/java/me/hufman/androidautoidrive/phoneui/viewmodels/CarDrivingStatsModel.kt#L33) is used to fetch the car's language at the start of the connection, to know what language the app modules should display in the car. Until the car language has been fetched, it does not continue onward to start up the rest of the app modules.

Because you mention that the background connectivity works, this should be working: As long as the car apps are connected and showing entries in the Online Services menu in the car, AAIdrive should be watching the car information and updating the remembered information (by default every 10 seconds, until you open the Car Information tab and it switches to realtime updates). Is that not what you experience?

**Reply by hufman** [reply]
I'm doing some closer investigating and there may be a tricky subtle bug in the Car Information handling. Can you try this test as a workaround:
1. Before you get into the car, force close the entire app and open the app, but make sure to not open the Car Information tab.
2. Then you can put away the phone and get into the car and drive as usual.
3. After you arrive and leave the car, check the Car Information tab and see if it looks correct.

When viewing the Car Information tab, it registers for faster updates, and then unregisters when the tab is closed. I had previously thought that the cached data registration should keep the car registration active, but some additional logging suggests that this may be buggy.

**Reply by hufman** [reply]
Thank you so much for reporting this bug! It's been hiding secretly for the last 4 months, since cf48383 :)
I just pushed up a new version, and the latest build now includes the fix.
Thank you so much!

**Reply by gab81** [reply]
hello Hufman, i didnt have time to reply your first message as i was sleeping lol - so you identified this was indeed an issue and there's a new version to test now, if i see this correctly, right?

So if i install this latest one, it should run all the background processes itself and also update the car's location when i arrive at destination without the need of me clicking the app, is that the idea?

Actually they had this in the "old" official mini app, with link to gmaps now the new one lacks so many features, hence that's why i use yours :)

thank you,
Gabrio

**Reply by gab81** [reply]
i just went to downloads page, it seems the same version, is it this one the newly refreshed one? https://github.com/BimmerGestalt/AAIdrive/releases/download/1.3/androidautoidrive-1.3-nomap-nonalytics-release.apk

**Reply by gab81** [reply]
or the sentry one?

**Reply by hufman** [reply]
Indeed, the filename for the automatic build is the same, but after installing you should see a new version number in the app details.
It should fix the automatic data updating, which was working previously as long as you never looked at the Car Info tab before driving heh.

**Comment by gab81** [comment]
sorry to be a pest, i just did that twice and the app version is still the same, perhaps you didn't upload ? :) it's still 1.3 28 Jul 2021 05.54.00 identical as before.

**Reply by hufman** [reply]
I apologize, I was sleepy and I didn't look closely at the link: That is indeed the link to the finalized 1.3 release.
When you mentioned you were using version 1.3.28, I thought you were using an automatic build which has a similar-looking version scheme (28 commits after 1.3, for example). Now I see that you are using 1.3 released on 28 July!

On the main Readme page, under Getting Started, there's a paragraph about the automatic nightly build. That has links to the [sentry](https://bimmergestalt.s3.amazonaws.com/aaidrive/builds/androidautoidrive-latest-main-nomap-sentry-release.apk) and [nonalytics](https://bimmergestalt.s3.amazonaws.com/aaidrive/builds/androidautoidrive-latest-main-nomap-nonalytics-release.apk) builds, and those links were what I referred to when I said the automatic build links don't change. Installing one of those will have the fix for you! I apologize for the confusion.

**Comment by gab81** [comment]
Yes so if it is july release is old? Shouldnt it say september like today?
😁

----------------------------

**Reply by hufman** [reply]
The new features [planned for version 1.3.1](https://github.com/BimmerGestalt/AAIdrive/issues?q=is%3Aissue+milestone%3A1.3.1) haven't been finished yet :)

**Comment by gab81** [comment]
So i did testing a little and im now out with the car.

I did what you said few posts ago:

Kill the app, then load it but not click car status. Then drive, park and
check app.

This saves last location perfect.

But if i kill app then load app once in the car go to car information then
open maps etc etc (not killing it), then once you arrive and park location
is NOT saved.

Now last location is 20 mins ago.

I re installed the most recent apk you have online on the no analytics
build, i thought you pushed live a fix just for this but doesnt look like.

Thanks for all your work it is really appreciated.

I love those oil and engine temp, having one of those little analog gauges
would be awesome too heheheh

Wish the mini app did all this!

Best regards,
Gabrio

----------------------------

**Reply by hufman** [reply]
I'm building separate apps to display gauges about car information, check out the cds_gauges app [over here](https://github.com/BimmerGestalt/IDriveConnectAddons)! Currently it's just an example app, for other people to build on and add their own gauges in the source code, but hopefully someone takes inspiration and builds out a more fully-featured customizable app!
![photo1627700452](https://user-images.githubusercontent.com/1592375/132604099-ead54c13-fbda-4c7a-b7e7-5da6800c1cb4.jpeg)

**Comment by gab81** [comment]
wow those are AWESOME!!!! 😀😁
